File: Programmers/hash_001.py
def solution(participant, completion):
    for p in participant:
        if participant.count(p) > completion.count(p):
            return p
    answer = ''
    return answer
    
# 정확성 100 + 효율성 50 코드 => brute force 이기 때문인것으로 보임
# list 는 


def solution(participant, completion):
    participant.sort()
    completion.sort()
    for p,c in zip(participant, completion):
        if p != c:
            return p
    return participant.pop()

# sort 함수는 O(nLogn) 의 속도를 가지기 때문에 위 보다 훨씬 나을 것으로 판단됨



import collections

def solution(participant, completion):
    answer = collections.Counter(participant) - collections.Counter(completion)
    return list(answer.keys())[0] # list 형태로 빼야하기 때문임

# collections 의 Counter 는 카운터에 가장 성능이 좋은 자료구조라고 함
# 해쉬로 구성되어있어서 접근이 빠르고 items() 에서 1번째, 즉 value 값에 따라서 정렬되어있다고 함
# 따라서 뺄셈을 진행했을때 1 인 값이 정답에 남게 되고 list로 반환하면 끝임 

File: Programmers/hash_001.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion){
        String answer = "";


       Map <String, Integer> hm = new HashMap<String, Integer>();

        for(String p_member : participant){
            if(hm.get(p_member) == null){
                hm.put(p_member, 1);
            }
            else{
                hm.put(p_member, hm.get(p_member) + 1);
            }
        }

        for(String c_mem : completion){
            hm.put(c_mem, hm.get(c_mem) - 1);
        }

        for(String key : hm.keySet()){
            if(hm.get(key) == 1) answer = key;
        }
        return answer;
    }
}

/*
    배운점
    1. HashMap 은 삽입시를 제외하고는 접근에서 o(1) 의 성능을 보여준다.
    2. 제너릭스를 사용하는 이유는 컴파일타임에 런타임 에러를 미리 잡기 위함이다(ex> 참조타입을 미리 설정해놓는다)
*/
